graphql_compiler/macros/macro_expansion.py

```diff
@@ -47,75 +67,105 @@ def _merge_selection_sets(selection_set_a, selection_set_b):
         field_a = selection_dict_a[field_name]
         field_b = selection_dict_b[field_name]
         if field_a.selection_set is not None or field_b.selection_set is not None:
-            raise GraphQLCompilationError('TODO')
+            raise GraphQLCompilationError(u'Macro expansion results in a query traversing the '
+                                          u'same edge {} twice, which is disallowed.'
+                                          .format(field_name))
 
-        merged_field = deepcopy(field_a)
-        merged_field.directives += field_b.directives
+        merged_field = copy(field_a)
+        merged_field.directives = field_a.directives + field_b.directives
         common_selection_dict[field_name] = merged_field
 
-    # Merge dicts
+    # Merge dicts, using common_selection_dict for keys present in both selection sets.
     merged_selection_dict = copy(selection_dict_a)
     merged_selection_dict.update(selection_dict_b)
-    merged_selection_dict.update(common_selection_dict)
+    merged_selection_dict.update(common_selection_dict)  # Overwrite keys in the intersection.
 
-    # Remove pro-forma fields if allowed
-    if len(merged_selection_dict) > 1:
+    # The macro or the user code could have an unused (pro-forma) field for the sake of not
+    # having an empty selection in a vertex field. We remove pro-forma fields if they are
+    # no longer necessary.
+    if len(merged_selection_dict) > 1:  # Otherwise we need a pro-forma field
         merged_selection_dict = {
             name: ast
             for name, ast in six.iteritems(merged_selection_dict)
             if ast.selection_set is not None or len(ast.directives) > 0
+            # If there's selections or directives under the field, it is not pro-forma.
         }
 
     # Get a deterministic ordering of the merged selections
-    selection_name_order = [
+    selection_name_order = list(chain((
         ast.name.value
         for ast in selection_set_a.selections
         if ast.name.value not in selection_dict_b
-    ] + [
+    ), (
         ast.name.value
         for ast in selection_set_b.selections
-    ]
+    )))
 
     # Make sure that all property fields come before all vertex fields.
-    return SelectionSet([
+    return SelectionSet(list(chain((
         merged_selection_dict[name]
         for name in selection_name_order
         if name in merged_selection_dict and merged_selection_dict[name].selection_set is None
-    ] + [
+    ), (
         merged_selection_dict[name]
         for name in selection_name_order
         if name in merged_selection_dict and merged_selection_dict[name].selection_set is not None
-    ])
+    ))))
```

Comment:
This is fairly repetitive. We already have a dependency on `funcy`, maybe `funcy.split` would help?

https://funcy.readthedocs.io/en/stable/seqs.html#lsplit

Reply:
Also, I realize I suggested this approach on the previous iteration -- in retrospect that wasn't a good idea, sorry about that.